### ISSUE TITLE
Segmentation fault with phyml-mpi if there are duplicate sequences

### ISSUE THREAD

**Issue by protist** [stephaneguindon/phyml#106]
## Summary
If there are duplicate sequences, mpi-enabled phyml crashes with a segmentation fault. Single-core phyml works fine.

## Steps to replicate
1. Create `test.phy` as follows
```
3 3
seq1      KDE
seq2      KKE
seq3      KDE
```
2. Build phyml with `--enable-mpi` as per [this package](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=phyml-mpi).
3. Run `mpirun -n 5 --use-hwthread-cpus phyml-mpi -i test.phy --datatype aa -b 10`
4. The following output is shown.

```
. Note: taxon 'seq1' is a duplicate of taxon 'seq3'.
...
. Final optimisation steps...
[<hostname>:24320] *** Process received signal ***
[<hostname>:24320] Signal: Segmentation fault (11)
[<hostname>:24320] Signal code: Address not mapped (1)
[<hostname>:24320] Failing at address: (nil)
[<hostname>:24322] *** Process received signal ***
[<hostname>:24322] Signal: Segmentation fault (11)
[<hostname>:24322] Signal code: Address not mapped (1)
[<hostname>:24322] Failing at address: (nil)


. Log likelihood of the current tree: -2125.189256.[<hostname>:24322] [ 0] /usr/lib/libpthread.so.0(+0x13cb0)[0x7f1b70bcdcb0]
[<hostname>:24322] [ 1] phyml-mpi(+0x13fd4a)[0x55775b13fd4a]
[<hostname>:24322] [ 2] [<hostname>:24320] [ 0] phyml-mpi(+0x3b56)[0x55775b003b56]
[<hostname>:24322] [ 3] /usr/lib/libc.so.6(__libc_start_main+0xf3)[0x7f1b70a1dee3]
[<hostname>:24322] [ 4] phyml-mpi(+0x447e)[0x55775b00447e]
[<hostname>:24322] *** End of error message ***
/usr/lib/libpthread.so.0(+0x13cb0)[0x7f2677e5fcb0]
[<hostname>:24320] [ 1] phyml-mpi(+0x13fd4a)[0x555927c5ad4a]
[<hostname>:24320] [ 2] phyml-mpi(+0x3b56)[0x555927b1eb56]
[<hostname>:24320] [ 3] /usr/lib/libc.so.6(__libc_start_main+0xf3)[0x7f2677cafee3]
[<hostname>:24320] [ 4] phyml-mpi(+0x447e)[0x555927b1f47e]
[<hostname>:24320] *** End of error message ***
[<hostname>:24318] *** Process received signal ***
[<hostname>:24318] Signal: Segmentation fault (11)
[<hostname>:24318] Signal code: Address not mapped (1)
[<hostname>:24318] Failing at address: (nil)
[<hostname>:24318] [ 0] /usr/lib/libpthread.so.0(+0x13cb0)[0x7fbc41c9fcb0]
[<hostname>:24318] [ 1] phyml-mpi(+0x13fd4a)[0x55b5a76acd4a]
[<hostname>:24318] [ 2] phyml-mpi(+0x3b56)[0x55b5a7570b56]
[<hostname>:24318] [ 3] /usr/lib/libc.so.6(__libc_start_main+0xf3)[0x7fbc41aefee3]
[<hostname>:24318] [ 4] phyml-mpi(+0x447e)[0x55b5a757147e]
[<hostname>:24318] *** End of error message ***
[<hostname>:24319] *** Process received signal ***
[<hostname>:24319] Signal: Segmentation fault (11)
[<hostname>:24319] Signal code: Address not mapped (1)
[<hostname>:24319] Failing at address: (nil)
[<hostname>:24321] *** Process received signal ***
[<hostname>:24321] Signal: Segmentation fault (11)
[<hostname>:24321] Signal code: Address not mapped (1)
[<hostname>:24321] Failing at address: (nil)
[<hostname>:24319] [ 0] [<hostname>:24321] [ 0] /usr/lib/libpthread.so.0(+0x13cb0)[0x7fc46b547cb0]
[<hostname>:24319] [ 1] /usr/lib/libpthread.so.0(+0x13cb0)[0x7f040d560cb0]
[<hostname>:24321] [ 1] phyml-mpi(+0x13fd4a)[0x558722c9ed4a]
[<hostname>:24321] [ 2] phyml-mpi(+0x3b56)[0x558722b62b56]
[<hostname>:24321] phyml-mpi(+0x13fd4a)[0x556a16a95d4a]
[<hostname>:24319] [ 2] phyml-mpi(+0x3b56)[0x556a16959b56]
[<hostname>:24319] [ 3] [ 3] /usr/lib/libc.so.6(__libc_start_main+0xf3)[0x7fc46b397ee3]
[<hostname>:24319] [ 4] /usr/lib/libc.so.6(__libc_start_main+0xf3)[0x7f040d3b0ee3]
[<hostname>:24321] [ 4] phyml-mpi(+0x447e)[0x558722b6347e]
phyml-mpi(+0x447e)[0x556a1695a47e]
[<hostname>:24319] *** End of error message ***
[<hostname>:24321] *** End of error message ***
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 0 with PID 0 on node <hostname> exited on signal 11 (Segmentation fault).
--------------------------------------------------------------------------
```
## Additional information
* It works if I run the same command with single-core phyml, i.e. building from [here](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=phyml), then running `phyml -i test.phy --datatype aa -b 10`.
* If I modify the duplicate sequence to make it unique, it works with phyml-mpi again.

I'm using the latest version of phyml 3.3.20190321.

**Comment by stephaneguindon** [comment]
Thanks. I'll look into that asap.

**Comment by protist** [comment]
No worries. TBH it's not a big deal once identified. It's easy enough to fix the file in advance, but it took me a while to work out why it was failing!

**Comment by stephaneguindon** [comment]
It's fixed in commit aa856487a55dea8db60c4ff6d78a7c11682afb65.
